include/ginkgo/core/matrix/scaled_permutation.hpp

```diff
@@ -0,0 +1,203 @@
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_MATRIX_SCALED_PERMUTATION_HPP_
+#define GKO_PUBLIC_CORE_MATRIX_SCALED_PERMUTATION_HPP_
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+/**
+ * ScaledPermutation is a matrix combining a permutation with scaling factors.
+ * It is a combination of Diagonal and Permutation, and can be read as
+ * $SP = P \pdot S$, i.e. the scaling gets applied before the permutation.
+ *
+ * @tparam IndexType  index type of permutation indices
+ * @tparam ValueType  value type of the scaling factors
+ *
+ * @ingroup permutation
+ * @ingroup mat_formats
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class ScaledPermutation
+    : public EnableLinOp<ScaledPermutation<ValueType, IndexType>>,
+      public WritableToMatrixData<ValueType, IndexType> {
+    friend class EnablePolymorphicObject<ScaledPermutation, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Returns a pointer to the scaling factors.
+     *
+     * @return the pointer to the scaling factors.
+     */
+    value_type* get_scale() noexcept { return scale_.get_data(); }
```

Comment:
On its own, scale is ambiguous, but together with `get_`, I think it's fairly unambiguous.

Reply:
I am also increasingly unhappy with `get_scale`, since it doesn't show that this is multiple values. I can't come up with a better name, so I'll use yours :)